[CMOB-1327] Feature: Create UI

tivanashvili opened the pull request:
UI part of one screen chatting app

Created one view, called ChatView which is instantiated twice in the VC, since we have two identical chat view on the screen.
Created separate component for switchButton which allows the user to set dark/light mode color.
Created separate component for textview, which dynamically increases its height up to five lines and after that is becomes scrollable.
Added UIVIew for yellow line which separates top and bottom chatView.
Added message bubbles based on figma

UI supports both the landscape mode and any other device
https://www.figma.com/file/UgTXNAXH2ScX6OHRREkSIZ/Designs-for-interns?type=design&node-id=45-147&t=kphGfsbEkq09yavZ-0

<img src="https://github.com/tivanashvili/Chat/assets/94744994/112d6de0-1d56-4320-b589-930cc0dee5ff" data-canonical-src="https://github.com/tivanashvili/Chat/assets/94744994/112d6de0-1d56-4320-b589-930cc0dee5ff" width="200" height="400"/>

<img src="https://github.com/tivanashvili/Chat/assets/94744994/f64d2b0f-4bd8-4a84-98b3-36153685bf1d" data-canonical-src="https://github.com/tivanashvili/Chat/assets/94744994/f64d2b0f-4bd8-4a84-98b3-36153685bf1d" width="200" height="400"/>

<img src="https://github.com/tivanashvili/Chat/assets/94744994/46f24954-1320-4cb1-bb3e-556e54ec3706" data-canonical-src="https://github.com/tivanashvili/Chat/assets/94744994/46f24954-1320-4cb1-bb3e-556e54ec3706" width="400" height="200"/>

<img src="https://github.com/tivanashvili/Chat/assets/94744994/d1e88403-9741-40b3-abd3-9f1b8d82496d" data-canonical-src="https://github.com/tivanashvili/Chat/assets/94744994/d1e88403-9741-40b3-abd3-9f1b8d82496d" width="400" height="200"/>

tivanashvili committed:
Remove extra comment

tivanashvili committed:
test commit

tivanashvili committed:
added design components

tivanashvili committed:
Added new components

Added dark/light mode switcher

tivanashvili committed:
Added UI Components

tivanashvili committed:
Added new components

tivanashvili committed:
Made changes according to review

MaksymTeslia commented on Chat App/Chat App/ChatComponent/ChatView.swift:
```diff
@@ -0,0 +1,103 @@
+
+final class ChatView: UIView {
+    
+    //MARK: variables
```
Better to declared marks like: 

`// MARK: Variables`

MaksymTeslia commented on Chat App/Chat App/ChatComponent/ChatView.swift:
```diff
@@ -0,0 +1,103 @@
+        tableView.register(MessageTableViewCell.self, forCellReuseIdentifier: "MessageTableViewCell")
+        tableView.backgroundColor = .red
+        return tableView
+    }()
```
That is cool to have everything settled at one place, but this way has a little tricky part. 

As you are setting data source in this block, you need a `tableView` property to be lazy var, which is actual variable. For the app to be as performant as possible, we need to declare all the properties as constants, which is in this case reachable by placing data source setting into a separate function and calling this function in `func setUp`.

PS. Please place `"MessageTableViewCell"` identifier into separate Constants enum as a static let.

MaksymTeslia commented on Chat App/Chat App/ChatComponent/ChatView.swift:
```diff
@@ -0,0 +1,103 @@
+    
+    private func setUpTableView() {
+        self.addSubview(tableView)
+    }
```
Please remove `self` keywords, cause we don't need to declare it explicitly.

MaksymTeslia commented on Chat App/Chat App/ChatComponent/ChatView.swift:
```diff
@@ -0,0 +1,103 @@
+struct Message {
+    let text: String
+    let date: Date
+}
```
Please place it into a separate file, it is a good practice to have all the models in dedicated files.

MaksymTeslia commented on Chat App/Chat App/ChatComponent/ChatView.swift:
```diff
@@ -0,0 +1,103 @@
+    }
+    
+    
+}
```
To mane spaces after function I guess )

MaksymTeslia commented on Chat App/Chat App/ChatTableView/MessageTableViewCell.swift:
```diff
@@ -0,0 +1,80 @@
+//
+import UIKit
+
+class MessageTableViewCell: UITableViewCell {
```
Should be final? 

Also, please add MARKs in this class.

MaksymTeslia commented on Chat App/Chat App/ChatTableView/MessageTableViewCell.swift:
```diff
@@ -0,0 +1,80 @@
+        label.textColor = .gray
+        label.font = UIFont.systemFont(ofSize: 12)
+        return label
+    }()
```
Don't need them to be lazy I think.

MaksymTeslia commented on Chat App/Chat App/ChatTableView/MessageTableViewCell.swift:
```diff
@@ -0,0 +1,80 @@
+    
+    private func setUpMessageLabel() {
+        contentView.addSubview(messageLabel)
+        messageLabel.numberOfLines = 0
```
Can be substituted by `.zero` I guess?

MaksymTeslia commented on Chat App/Chat App/ChatTableView/MessageTableViewCell.swift:
```diff
@@ -0,0 +1,80 @@
+        let formatter = DateFormatter()
+        formatter.dateFormat = "HH:mm"
+        return formatter.string(from: date)
+    }
```
Nice idea, but I would make `dateFormat` an input value of this function, so if "business requirements" change, it will be easy to modify the logic. Also, it makes it more reusable.

MaksymTeslia commented on Chat App/Chat App/MessageManager/Messages.swift:
```diff
@@ -0,0 +1,32 @@
+
+//import UIKit
+//
+//class Messages {
```
Usually, commented code isn't kept in production.

MaksymTeslia commented on Chat App/Chat App/Switcher/SwitcherView.swift:
```diff
@@ -0,0 +1,61 @@
+    func didToggleDayLightSwitch(isOn: Bool)
+}
+
+class DayLightSwitch: UIButton {
```
Should be final?

MaksymTeslia commented on Chat App/Chat App/Switcher/SwitcherView.swift:
```diff
@@ -0,0 +1,61 @@
+
+import UIKit
+
+protocol DayLightSwitchDelegate: AnyObject {
```
Please create a separate 'Protocols' folder in 'SwitcherView' and place this protocol to a separate file there. It is a good practice to have everything managed out by "logic".

MaksymTeslia commented on Chat App/Chat App/Switcher/SwitcherView.swift:
```diff
@@ -0,0 +1,61 @@
+    private let button = UIButton(type: .system)
+    private let dayImage = UIImage(named: Constants.ImageNames.lightModeSwitchButton)
+    private let nightImage = UIImage(named: Constants.ImageNames.darkModeSwitchButton)
+    private var isOn:Bool = false
```
Please add space before "Bool"

MaksymTeslia commented on Chat App/Chat App/Switcher/SwitcherView.swift:
```diff
@@ -0,0 +1,61 @@
+    private let nightImage = UIImage(named: Constants.ImageNames.darkModeSwitchButton)
+    private var isOn:Bool = false
+    
+    weak var delegate: DayLightSwitchDelegate!
```
Please don't use force unwrap, it is not safe, because if you, for example, forgot to set the delegate, the app will crash.

MaksymTeslia commented on Chat App/Chat App/Switcher/SwitcherView.swift:
```diff
@@ -0,0 +1,61 @@
+    }
+    
+    @objc private func buttonTapped() {
+        isOn = !isOn
```
Such a declaration can be substituted as `isOn.toggle()`, this is a standard Apple SDK function.

MaksymTeslia commented on Chat App/Chat App/Switcher/SwitcherViewComponentModels.swift:
```diff
@@ -0,0 +1,10 @@
+//  Created by tornike on 29.04.23.
+//
+
+//struct SwitcherViewComponentModels {
+//    let switcherImageName: String
+//}
```
Here commented code as well, if you don't need it, please remove.

MaksymTeslia commented on Chat App/Chat App/TextInputComponent/TextInputComponentView.swift:
```diff
@@ -0,0 +1,154 @@
+        containerView.layer.cornerRadius = 28
+        containerView.translatesAutoresizingMaskIntoConstraints = false
+        
+    }
```
Extra space on line 44.

MaksymTeslia commented on Chat App/Chat App/TextInputComponent/TextInputComponentView.swift:
```diff
@@ -0,0 +1,154 @@
+    }
+}
+    
+    extension TextInputComponentView: UITextViewDelegate{
```
Something happened with alignment, please fix (you can highlight all the code and press control + I, it will happen automatically)

MaksymTeslia commented on Chat App/Chat App/TextInputComponent/TextInputComponentView.swift:
```diff
@@ -0,0 +1,154 @@
+            containerView.layoutIfNeeded()
+        }
+}
+extension UITextView {
```
Please add space between them.

MaksymTeslia commented on Chat App/Chat App/ViewController.swift:
```diff
@@ -9,11 +9,69 @@ import UIKit
+        view.addSubview(topChatView)
+        view.addSubview(separatorView)
+        view.addSubview(bottomChatView)
+        view.addSubview(switchButton)
```
If you add such a function as a UIView extension: 

`func addSubviews(_ views: UIView...)`
`views.forEach({ addSubview($0) })`
`}`

you will be able to make this assignment single line, like:

`view.addSubviews(topChatView, separatorView, bottomChatView, switchButton)`


Which is more beautiful I think )

MaksymTeslia commented on Chat App/Chat App/ViewController.swift:
```diff
@@ -9,11 +9,69 @@ import UIKit
-
+    // MARK: layout constraints
+    private func setUpConstraintsForSubviews() {
+        topChatView.translatesAutoresizingMaskIntoConstraints = false
+        separatorView.translatesAutoresizingMaskIntoConstraints = false
+        bottomChatView.translatesAutoresizingMaskIntoConstraints = false
+        switchButton.translatesAutoresizingMaskIntoConstraints = false
```
It is a good practice to place any repetitive assignment to a loop, like this:

`[topChatView, separatorView, bottomChatView, switchButton].forEach({`
`$0.translatesAutoresizingMaskIntoConstraints = false`
`})`

MaksymTeslia commented on Chat App/Chat App/ViewController.swift:
```diff
@@ -9,11 +9,69 @@ import UIKit
+{
+    
+    func didToggleDayLightSwitch(isOn: Bool) {
+        if isOn {
+            // Dark mode is on
+            topChatView.backgroundColor = Constants.ChatAppColors.darkModeColor
+            bottomChatView.backgroundColor = Constants.ChatAppColors.darkModeColor
+            separatorView.backgroundColor = Constants.ChatAppColors.darkModeColor
+        } else {
+            // Day mode is on
+            topChatView.backgroundColor = .clear
+            bottomChatView.backgroundColor = .clear
+            separatorView.backgroundColor = .clear
+        }
```
Ternary expression in case of simple if-else statement is a good choice, for example: 

`topChatView.backgroundColor = isOn ? Constants.ChatAppColors.darkModeColor : .clear`
ect...

MaksymTeslia commented on Chat App/Chat App/ChatComponent/ChatView.swift:
```diff
@@ -0,0 +1,103 @@
+    //MARK: variables
+    
+    private let textInputComponentView = TextInputComponentView()
+    private var messages = [
+        Message(text: "Hello", date: Date()),
+        Message(text: "How are you?", date: Date()),
+        Message(text: "I'm doing well, thanks!", date: Date())
+    ]
```
Have not found any assignment to this array, so should we make it a constant maybe ?

MaksymTeslia commented on Chat App/Chat App/ChatComponent/ChatView.swift:
```diff
@@ -0,0 +1,103 @@
+    }
+    
+    func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
+        let cell = tableView.dequeueReusableCell(withIdentifier: "MessageTableViewCell", for: indexPath) as! MessageTableViewCell
```
When you place `"MessageTableViewCell"` identifier to Constants, don't forget to replace it here as well. 

PS. Better to avoid force unwrap and use optional binding with 'as?' keyword.

MaksymTeslia commented on Chat App/Chat App/Constants/CommonConstatns.swift:
```diff
@@ -0,0 +1,85 @@
+        static let placeHolder = "დაწერე შეტყობინება..."
+    }
+    enum ImageNames {
+        static let SendButton = "sendButtonImage"
```
Please change constant name to lowercased.

MaksymTeslia commented on Chat App/Chat App/TextInputComponent/TextInputComponentView.swift:
```diff
@@ -0,0 +1,154 @@
+    
+    private let button = UIButton()
+    
+    private var containerViewHeightConstraint: NSLayoutConstraint!
```
Here I'd avoid force-unwrap as well.

tivanashvili committed:
Made improvements to UI after the review

lozovyi-andersen commented on Chat App/Chat App/ChatTableView/MessageCellConstants.swift:
```diff
@@ -0,0 +1,21 @@
+//
+
+import UIKit
+
+enum Constants {
+    
```
should this Constants enum be global or be in MessageTableViewCell extension?

lozovyi-andersen commented on Chat App/Chat App/Switcher/SwitcherView.swift:
```diff
@@ -0,0 +1,54 @@
+//
+
+import UIKit
+
+enum Button {
+    case light
```
we have Button in swiftui, and bad enum naming in general. Pls name it more self-explanatory

lozovyi-andersen commented on Chat App/Chat App/TextInputComponent/TextInputComponentView.swift:
```diff
@@ -0,0 +1,145 @@
+    
+    func calculateMaxLines() -> Int {
+        let maxSize = CGSize(width: self.frame.width, height: CGFloat(Float.infinity))
+        let font = self.font ?? .systemFont(ofSize: 16)
```
we can move font size value to Constants as well

lozovyi-andersen commented on Chat App/Chat App/TextInputComponent/TextInputComponentView.swift:
```diff
@@ -0,0 +1,145 @@
+        textView.textColor = Constants.TextView.placeholderColor
+        textView.delegate = self
+        textView.contentOffset = CGPoint(x: 0, y: (textView.contentSize.height - textView.frame.height) / 2)
+        textView.textContainerInset = UIEdgeInsets(top: 19, left: 22, bottom: 19, right: 0)
+        textView.font = UIFont.systemFont(ofSize: 16)
```
we can move this values to Constants

tivanashvili replied:
i moved font to constant but if i move insets to constant also it will be textView.textContainerInset = UIEdgeInsets(top: Constants.TextViewInsets.top, left: Constants.TextViewInsets.left, bottom: Constants.TextViewInsets.bottom, right: Constants.TextViewInsets.right) like that, is not it bad for readability? or it must be like that anyway?

lozovyi-andersen replied:
you can add `static let textInsets = UIEdgeInsets(top: 19, left: 22, bottom: 19, right: 0)` in Constants

MaksymTeslia commented on Chat App/Chat App/ChatComponent/ChatView.swift:
```diff
@@ -0,0 +1,101 @@
+        Message(text: "Hello", date: Date()),
+        Message(text: "How are you?", date: Date()),
+        Message(text: "I'm doing well, thanks!", date: Date())
+    ]
+    
+    
+    private let tableView: UITableView = {
```
Pls remove extra space.

MaksymTeslia commented on Chat App/Chat App/ChatComponent/ChatView.swift:
```diff
@@ -0,0 +1,101 @@
+        cell.configure(with: message)
+        return cell
+    }
+}
+
+
+
+
```
Pls remove extra spaces.

MaksymTeslia commented on Chat App/Chat App/Constants/ViewControllerConstants.swift:
```diff
@@ -0,0 +1,36 @@
+            static let bottom: CGFloat = -30
+        }
+    }
+}
+
+
+
+
```
Pls remove extra spaces.

MaksymTeslia commented on Chat App/Chat App/Switcher/SwitcherView.swift:
```diff
@@ -0,0 +1,54 @@
+        self.isSelected.toggle()
+        delegate?.didToggleDayLightSwitch(with: self.isSelected ? .light : .dark)
+    }
+}
+
+
+
+
+
```
Pls remove extra spaces.

MaksymTeslia commented on Chat App/Chat App/TextInputComponent/TextInputComponentView.swift:
```diff
@@ -0,0 +1,145 @@
+        let linesRoundedUp = Int(ceil(textSize.height / charSize))
+        return linesRoundedUp
+    }
+}
+
+
+
+
```
Pls remove extra spaces.

MaksymTeslia commented on Chat App/Chat App/ViewController.swift:
```diff
@@ -8,12 +8,69 @@
 import UIKit
 
 class ViewController: UIViewController {
```
class should be 'final' I think

tivanashvili committed:
Made final changes

MaksymTeslia commented on Chat App/Chat App/Extensions/TableViewCell+Extensions.swift:
```diff
@@ -0,0 +1,17 @@
+
+import Foundation
+
+extension Date {
+        func formatDate() -> String {
+        let formatter = DateFormatter()
+        formatter.dateFormat = "HH:mm"
+        return formatter.string(from: self)
+    }
+}
```
Something happened to tabulation here. 

Also, please make `dateFormat` an input value for this function, this way function will be much more reusable.

tivanashvili replied:
you mean like this? extension Date {
    func formatDate(dateFormat: String) -> String {
        let formatter = DateFormatter()
        formatter.dateFormat = dateFormat
        return formatter.string(from: self)
    }
}

MaksymTeslia replied:
Yeap, absolutely) 

This way you will be able to call it from different classes and get strings of different formats if you need.

MaksymTeslia commented on Chat App/Chat App/ChatTableView/MessageTableViewCell.swift:
```diff
@@ -0,0 +1,117 @@
+        setUpLayoutConstraints()
+    }
+    
+    private func setUpChatBubbles() {
+        contentView.addSubview(chatBubble)
+        contentView.addSubview(middleBubble)
+        contentView.addSubview(smallestBubble)
+        chatBubble.translatesAutoresizingMaskIntoConstraints = false
+        middleBubble.translatesAutoresizingMaskIntoConstraints = false
+        smallestBubble.translatesAutoresizingMaskIntoConstraints = false
+    }
```
This way will be better I think: 

`[chatBubble, middleBubble, smallestBubble].forEach({`
`contentView.addSubivew($0)`
`$0.translatesAutoresizingMaskIntoConstraints = false`
`})`

LiziChichua commented on Chat App/Chat App/Switcher/SwitcherView.swift:
```diff
@@ -0,0 +1,49 @@
+final class DayLightSwitch: UIButton {
+    
+    // MARK: Components
+    
```
მარკების მერე 1 ხაზს ნუ ტოვებ, სხვაგანაც იგივე სტილით გაქვს და გაითვალისწინე ყველგან

LiziChichua commented on Chat App/Chat App/Switcher/SwitcherView.swift:
```diff
@@ -0,0 +1,49 @@
+}
+
+protocol DayLightSwitchDelegate: AnyObject {
+    func didToggleDayLightSwitch(with state: BackgroundMode)
```
ძალიან გრძელი სახელია და წასაკითხად რთულია. შეამოკლე და დაარქვი მაგალითად `didToggleSwitch(with state: BackgroundMode)`

LiziChichua commented on Chat App/Chat App/Switcher/SwitcherView.swift:
```diff
@@ -0,0 +1,49 @@
+    
+    @objc private func buttonTapped() {
+        self.isSelected.toggle()
+        delegate?.didToggleDayLightSwitch(with: self.isSelected ? .light : .dark)
```
`self.` ის დაწერა არ არის საჭირო არც 46 და არც 47 ხაზზე.

LiziChichua commented on Chat App/Chat App/ChatComponent/ChatView.swift:
```diff
@@ -0,0 +1,98 @@
+    
+    // MARK: Components
+    private let textInputComponentView = TextInputComponentView()
+    private let messages = [
```
ჯერ კომპონენტები უნდა დაალაგო და მერე ფროფერთები. `textInputComponentView` კომპონენტია, მერე შუაში `messages` გაქვს ჩაწერილი, რაც არ არის კომპონენტი და მაგის ქვევით კიდე `tableView` რომელიც კომპონენტია. ჯერ დაწერე  `textInputComponentView & tableView` მერე დაადე ფროფერთების მარკი და მარკის ქვევით დაწერე `messages`

LiziChichua commented on Chat App/Chat App/ChatComponent/ChatView.swift:
```diff
@@ -0,0 +1,98 @@
+    
+    func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
+        guard let cell = tableView.dequeueReusableCell(withIdentifier: Constants.CellConstants.cellIdentifier, for: indexPath) as? MessageTableViewCell else {
+            fatalError("Unable to dequeue MessageTableViewCell")
```
რევიუზე გახსოვს რა გითხარით?

tivanashvili replied:
force unwrap ხო?

LiziChichua commented on Chat App/Chat App/ChatTableView/MessageTableViewCell.swift:
```diff
@@ -0,0 +1,117 @@
+final class MessageTableViewCell: UITableViewCell {
+    
+    // MARK: Components
+    private let messageLabel: UILabel = {
```
კომპონენტებში სულ ციფრები გაქ - გაიტანე კონსტანტებში

LiziChichua commented on Chat App/Chat App/ChatTableView/MessageTableViewCell.swift:
```diff
@@ -0,0 +1,117 @@
+    
+    private func setUpMessageLabel() {
+        chatBubble.addSubview(messageLabel)
+        messageLabel.translatesAutoresizingMaskIntoConstraints = false
```
`translatesAutoresizingMaskIntoConstraints` შეგიძლია მე-12 ხაზზე რომ `messageLabel` ის ქლოჟერი გაქვს მანდ გაუწერო. უფრო სწორია მანდ რომ გაეწერო და ეს ფუნქცია საერთოდ წაშალო. საბვიუების დასამატებლად ცალკე ფუნქცია დაწერე და იქ დაამატე `messageLabel` საბვიუდ და იგივე გააკეთე `dateLabel` ზეც.

LiziChichua commented on Chat App/Chat App/ChatTableView/MessageTableViewCell.swift:
```diff
@@ -0,0 +1,117 @@
+    
+    private func setUpDateLabel() {
+        contentView.addSubview(dateLabel)
+        dateLabel.translatesAutoresizingMaskIntoConstraints = false
```
იგივე გააკეთე რაც `setUpMessageLabel` ის ფუნქციაში დაგიწერე

LiziChichua commented on Chat App/Chat App/TextInputComponent/TextInputComponentView.swift:
```diff
@@ -0,0 +1,144 @@
+    }
+    
+    private func setUpTextView() {
+        textView.text = Constants.TextView.placeHolder
```
ეს მთლიანი სეთაფი `textView` ს ქლოჟერში გაიტანო სჯობს. ეს ფუნქციაც აღარ დაგჭირდება. როგორც `MessageTableViewCell` ში გაქ კომპონენტები ქლოჟერით გაკეთებული მაგ სტილს ვგულისხმობ.

LiziChichua commented on Chat App/Chat App/TextInputComponent/TextInputComponentView.swift:
```diff
@@ -0,0 +1,144 @@
+    
+    private func setUpButton() {
+        button.setImage(UIImage(named: Constants.TextView.sendButton), for: .normal)
+        button.translatesAutoresizingMaskIntoConstraints = false
```
აქაც იგივე

LiziChichua commented on Chat App/Chat App/TextInputComponent/TextInputComponentView.swift:
```diff
@@ -0,0 +1,144 @@
+    
+    private func updateTextViewHeight() {
+        let maxLines = textView.calculateMaxLines()
+        if maxLines > 5 {
```
ეს 5 კონსტანტებში უნდა იყოს + გავტესტე და ეს ლოგიკები კარგად არ მუშაობს, ვიზუალური ბაგია

LiziChichua commented on Chat App/Chat App/TextInputComponent/TextInputComponentView.swift:
```diff
@@ -0,0 +1,144 @@
+        }
+    }
+    
+    private func setUpBubbles() {
```
ეს ფუნქცია აქ რა შუაშია? წაშალე

LiziChichua commented on Chat App/Chat App/TextInputComponent/TextInputComponentView.swift:
```diff
@@ -0,0 +1,144 @@
+    }
+}
+
+extension UITextView {
```
ეს ექსთენშენი აქ არ უნდა იყოს. `Extensions` ფოლდერში ახალი ფაილი დაამატე `UITextView+Extensions` და იქ გაიტანე

LiziChichua commented on Chat App/Chat App/ViewController.swift:
```diff
@@ -7,13 +7,67 @@
+    func didToggleDayLightSwitch(with state: BackgroundMode) {
+        switch state {
+        case .light:
+            [topChatView, bottomChatView, view].forEach { $0.backgroundColor = Constants.ViewColor.darkModeColor }
```
68-ე და 70 ხაზზე გაქვს იდენტური ლოგიკა, შესაბამისად ორჯერ რომ წერ გამოგდის კოდის დუბლირება. ფუნქციაში გაიტანე და პარამეტრად გადააწოდე `state: BackgroundMode`. ისე, ცოტა ლოგიკა დამაბნეველია... `.light` ის დროს `darkModeColor` ს უსეტავ და `.dark` ის დროს  `. white` ს. ესე რატომ ხდება?

LiziChichua reviewed (changes requested):
კოდში პატარა კომენტარები დაგიტოვე, რომელიც მარტივად გასწორდება. ცუდი ის არის რომ `TextInputComponentView` ს სქროლვას აქვს ვიზუალური ბაგი. ამ ყველაფერს რომ გაასწორებ მერე კიდე გადავხედავ.

tivanashvili committed:
Made Final changes according to review and PR

LiziChichua commented on Chat App/Chat App/Extensions/TableViewCell+Extensions.swift:
```diff
@@ -0,0 +1,17 @@
+//
+//  TableViewCell+Extensions.swift
```
ფაილს ქვია `TableViewCell+Extensions` არადა ამ დროს `Date` ის ექსთენშენი გაქვს. შეცვალე ფაილის სახელი და დაარქვი `Date+Extensions`

LiziChichua reviewed (changes requested):
1 - ყველა კლასი `command + a` `control + i` ს დააჭირე რომ გასწორდეს ინდენტაცია
2 - გაკლია sender ის cell ი

tivanashvili committed:
Made improvements on UI

tivanashvili committed:
Fixed spaces, Changed name of file according to review

shineRR commented on Chat App/Chat App/ChatTableView/MessageTableViewCell.swift:
```diff
@@ -0,0 +1,206 @@
+        ])
+    }
+    
+    func configure(with message: String, indexPath: IndexPath) {
```
You should pass the model for presentation to your cell.
For instance:
```suggestion
    enum BubblePosition { case left, right }
    struct BubbleMessage { 
        let position: BubblePosition
        let message: String
    }
```

tivanashvili committed:
Add dummy data and BubblePosition enum

This commit adds dummy data for messages and introduces a BubblePosition enum to represent the position of each message in the chat view. The dummy data provides example messages for testing and demonstration purposes. The BubblePosition enum allows distinguishing between left and right positions for message bubbles, enabling proper layout configuration in the chat view.

LiziChichua commented on Chat App/Chat App/MessageManager/Messages.swift:
```diff
@@ -0,0 +1,19 @@
+}
+
+struct Messages {
+    let userID: BubblePosition
```
`userID` ის ნაცვლად `bubblePosition` დაარქვი

LiziChichua replied:
ჯერ ეს დაგიწერე და მერე შევხედე დანარჩენ ლოგიკას...

`let userID: Int`

წესით ესე დაგჭირდება

LiziChichua commented on Chat App/Chat App/MessageManager/Messages.swift:
```diff
@@ -0,0 +1,19 @@
+    case right
+}
+
+struct Messages {
```
ეს არის მესიჯის მოდელი ამიტომ მხოლობით რიცხვში უნდა ერქვას სახელი. ზოგადად მოდელები ყოველთვის მხოლობით რიცხვშია. `Messages` ის ნაცვლად `Message` უნდა იყოს

LiziChichua commented on Chat App/Chat App/ChatComponent/ChatView.swift:
```diff
@@ -0,0 +1,108 @@
+            cell.setUpSenderBubbleColor()
+        }
+
+        cell.configure(with: message)
```
96-102 ხაზები წაშალე და დაწერე ესეთი სტილით:

`cell.configure(with: message, position: message.userID == 1 ? .left : .right)`

`userID` არ უნდა იყოს `BubblePosition` ტიპის. დანარჩენი შენით მოიფიქრე ესე როგორ უნდა ჩაწერო. მინიშნება მოგეცი უკვე :))

tivanashvili replied:
ანუ configure მეთოდში შევიტანო ბაბლის პოზიციის დახატვის ლოგიკა?

tivanashvili committed:
Refactor

tivanashvili committed:
Bubble position refactor

LiziChichua commented on Chat App/Chat App/ChatTableView/MessageTableViewCell.swift:
```diff
@@ -0,0 +1,208 @@
+//        Message(text: "Hello", date: Date()),
+//        Message(text: "How are you?", date: Date()),
+//        Message(text: "I'm doing well, thanks!", date: Date())
+//    ]
```
1-6 ხაზები წაშალე

LiziChichua commented on Chat App/Chat App/ChatTableView/MessageTableViewCell.swift:
```diff
@@ -0,0 +1,208 @@
+        dateLabel.text = message.date.formatDate(dateFormat: Constants.DateConfigure.date)
+        
+        if bubblePosition == BubblePosition.left {
+            setUpLayoutConstraints()
```
უფრო კითხვადი რომ იყოს `setUpLayoutConstraints()` ის მაგივრად setUpReceiverConstraints() დაარქვი და ანალოგიურად სენდერზეც

LiziChichua commented on Chat App/Chat App/ChatTableView/MessageTableViewCell.swift:
```diff
@@ -0,0 +1,208 @@
+        messageLabel.text = message.message
+        dateLabel.text = message.date.formatDate(dateFormat: Constants.DateConfigure.date)
+        
+        if bubblePosition == BubblePosition.left {
```
swifch bubblePosition დაწერე. ენამის დროს if else ებს არ ვიყენებთ

LiziChichua commented on Chat App/Chat App/ChatTableView/MessageTableViewCell.swift:
```diff
@@ -0,0 +1,208 @@
+        setUpMessageLabel()
+    }
+    
+    func setUpLayoutConstraints() {
```
private?

LiziChichua commented on Chat App/Chat App/ChatTableView/MessageTableViewCell.swift:
```diff
@@ -0,0 +1,208 @@
+        contentView.addSubview(middleBubble)
+        contentView.addSubview(smallestBubble)
+        contentView.addSubview(dateLabel)
+    }
```
ენტერი დაარტყი

LiziChichua commented on Chat App/Chat App/ChatTableView/MessageTableViewCell.swift:
```diff
@@ -0,0 +1,208 @@
+        ])
+    }
+    
+    func setUpSenderBubbleLayoutConstraints() {
```
private?

LiziChichua commented on Chat App/Chat App/ChatTableView/MessageTableViewCell.swift:
```diff
@@ -0,0 +1,208 @@
+        setUpRightDateLabelConstraints()
+    }
+    
+    func setUpSenderBubbleColor() {
```
აქაც private

LiziChichua commented on Chat App/Chat App/ChatComponent/ChatView.swift:
```diff
@@ -0,0 +1,100 @@
+        setUpChatView()
+    }
+    
+    private func setUpChatView() {
```
ეს ფუნქცია რას ემსახურება ვერ მივხვდი... გჭირდება საერთოდ?

tivanashvili committed:
refactor

tivanashvili committed:
Added TapGestureRecognizer, refactored according to review
